### ISSUE TITLE
About Scannet Datas

### ISSUE THREAD

**Issue by QingWindIsStillTheWind** [zurutech/3DBoNet2#1]
According to the paper, the author said that you used 18 categories in the scan data set for testing. During training, do you take other categories other than 18 categories as the background for training? Or is it that only 18 classes of point clouds are selected for training?

**Comment by lucagrementieri** [comment]
During training (and during testing too) we assign the other categories to background.

In fact:
- we assign to such point dummy instance and segmentation labels equal to -1 https://github.com/zurutech/3DBoNet2/blob/63c162573fc455306b3c985c8f9d6d352ebf54bd/src/bonet2/datasets/scannet/scannet.py#L435
- we do not consider bounding boxes of instances with label -1 https://github.com/zurutech/3DBoNet2/blob/63c162573fc455306b3c985c8f9d6d352ebf54bd/src/bonet2/datasets/cloud_blocks_dataset.py#L232
-  we do not consider the points with segmentation label -1 in the calculation of the cross-entropy loss https://github.com/zurutech/3DBoNet2/blob/63c162573fc455306b3c985c8f9d6d352ebf54bd/src/bonet2/losses/masked_categorical_crossentropy.py#L19

The model has to learn how to manage objects from unknown categories because at test time it has to recognise them.
The removal of points from unknown categories during training would be detrimental to performances, because it will create a gap between training and test.

**Comment by QingWindIsStillTheWind** [comment]
[3DBoNet2/src/bonet2/datasets/scannet/scannet.py]:
if label != -1:
                object_id = instance["objectId"]
                segments = instance["segments"]
                object2segments[object_id] = segments
                label2segments.setdefault(label, []).extend(segments)
Here we judge whether the semantic tag is the purpose of - 1, and whether to eliminate the point clouds outside 18 categories? In other words, according to your answer, does it mean that the semantic tags of categories other than 18 categories are treated as - 1 and used for training together?

**Comment by QingWindIsStillTheWind** [comment]
[3DBoNet2/src/bonet2/datasets/scannet/scannet.py]:
if label != -1:
object_id = instance["objectId"]
segments = instance["segments"]
object2segments[object_id] = segments
label2segments.setdefault(label, []).extend(segments)
Here we judge whether the semantic tag is the purpose of - 1, and whether to eliminate the point clouds outside 18 categories? In other words, according to your answer, does it mean that the semantic tags of categories other than 18 categories are treated as - 1 and used for training together?

**Comment by QingWindIsStillTheWind** [comment]
[3DBoNet2/src/bonet2/datasets/scannet/scannet.py]:

if label != -1:
object_id = instance["objectId"]
segments = instance["segments"]
object2segments[object_id] = segments
label2segments.setdefault(label, []).extend(segments)

Here we judge whether the semantic tag is the purpose of - 1, and whether to eliminate the point clouds outside 18 categories? In other words, according to your answer, does it mean that the semantic tags of categories other than 18 categories are treated as - 1 and used for training together?

**Comment by lucagrementieri** [comment]
The code you pasted, comes from `read_aggregation` that uses labels already processed to assign the label -1 to other categories.
https://github.com/zurutech/3DBoNet2/blob/63c162573fc455306b3c985c8f9d6d352ebf54bd/src/bonet2/datasets/scannet/scannet.py#L511

The processing of labels happens in `read_label_mapping`.
https://github.com/zurutech/3DBoNet2/blob/63c162573fc455306b3c985c8f9d6d352ebf54bd/src/bonet2/datasets/scannet/scannet.py#L450

> In other words, according to your answer, does it mean that the semantic tags of categories other than 18 categories are treated as - 1 and used for training together?

Yes, the points from categories other than the 18 selected categories are assigned to a -1 label and used in training accordingly.

**Comment by QingWindIsStillTheWind** [comment]
https://github.com/Yang7879/3D-BoNet/issues/21

Have you read the answer of 3dbonet author? In his reply, he said that you can eliminate other categories you are not interested in during training and then carry out training. What do you think?

**Comment by lucagrementieri** [comment]
Our results on ScanNet were very bad, so maybe that's a useful trick. Anyhow I don't understand how the model can obtain good results at test time, when it sees a point distribution very different from training data.

**Comment by QingWindIsStillTheWind** [comment]
The 3DBoNet source code version you uploaded is based on tensorflow2. Why is it so different from the source code test results provided by the original author?

**Comment by lucagrementieri** [comment]
Our code is based on the original source code provided, we manually converted the TF1 code to TF2 and used an object-oriented design similar to the PyTorch style.

The fact is that the original published source code only manages an already processed version of S3DIS and no pre-processing code is directly available. We have succeeded in reproducing the results on S3DIS, but ScanNet is a much larger dataset so the pre-processing has to be different.
We have designed a preprocessing pipeline on ScanNet based solely on common practices and the few suggestions given by the authors in the original repository, but probably we have missed some important parts that weren't explained anywhere.
